### ISSUE TITLE
Typescript - Interface injecting

### ISSUE THREAD

**Issue by jezikk** [kamilmysliwiec/nest#43]
Hi,

is it possible to inject a component based on its interface? 

Let's assume the following component:

`class UserRepository implements IUserRepository`

This component is injected into the controller by interface from which is implemented:

```
export class UserController {
    constructor(private userService: IUserRepository) {}
```
It would be great to be able to set dependencies as follows:

```
@Module({
    controllers: [ UsersController ],
    components: [
        { provide: IUserRepository, useClass: UserRepository }
    ],
})
```

Thanks

**Comment by zachgrayio** [comment]
I've wondered the same thing - would definitely be a useful feature. 

So far I've been injecting the concrete classes and just storing them in an instance var typed as the interface to at least keep things decoupled as much as possible.

**Comment by JulianBiermann** [comment]
Using Interfaces as value types is not possible as they only exist during development. After transpiling interfaces no longer exist resulting in empty object values. There is a solution for your problem though using string keys as provide values and the inject decorator:

```typescript
@Module({
    controllers: [ UsersController ],
    components: [
        { provide: 'IUserRepository', useClass: UserRepository }
    ],
})
```

```typescript
export class UserController {
    constructor(@Inject('IUserRepository') private userService: IUserRepository) {}
```

**Comment by kamilmysliwiec** [comment]
Hi @jezikk 
As @JulianBiermann said - it is not possible. TypeScript types disappear after compilation. They are not a part of the language.

**Comment by zachgrayio** [comment]
How would one export a string keyed component? 

```

**Comment by kamilmysliwiec** [comment]
Hi @zachgrayio,

It was an issue of the previous version. The bug is already resolved, just update package (`1.0.6`)
